### PR TITLE
Feature/nautilus final

### PR DESCRIPTION
**Pull request by Jammy2211**
Final features for nautilus: visualization and MPI.

### CODE REVIEW

File: autofit/non_linear/parallel/sneaky.py [hunk 1]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+    """
+
+
+def initializer(
+        fitness,
+        prior_transform,
+        fitness_args,
+        fitness_kwargs,
+        prior_transform_args,
+        prior_transform_kwargs
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:366]: Type annotations

File: autofit/non_linear/parallel/sneaky.py [hunk 2]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+
+
+def prior_transform_cache(x):
+    """
+    Prior transform call
+    """
+    return FunctionCache.prior_transform(x, *FunctionCache.prior_transform_args,
+                                         **FunctionCache.prior_transform_kwargs)
+
+class SneakierPool:
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:395]: lol

File: autofit/non_linear/parallel/sneaky.py [hunk 3]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+    def __init__(
+            self,
+            processes: int,
+            fitness: Callable,
+            prior_transform: Optional[Callable] = None,
+            fitness_args: Optional[Iterable] = None,
+            fitness_kwargs: Optional[dict] = None,
+            prior_transform_args: Optional[Iterable] = None,
+            prior_transform_kwargs: Optional[dict] = None,
+    ):
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:405]: Please add docs

File: autofit/non_linear/parallel/sneaky.py [hunk 4]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+
+    def is_master(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            return_value = self.comm.rank == 0
+        else:
+            return_value = True
+        
+        return return_value
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:444]:
```python
return not is_mpi or self.comm.rank == 0

File: autofit/non_linear/parallel/sneaky.py [hunk 5]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+
+    def wait(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            self.pool.wait()
+        else:
+            pass
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:452]: This pass statement does not do anything

File: autofit/non_linear/parallel/sneaky.py [hunk 6]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+        """
+
+        use_mpi = self.check_if_mpi()
+
+        if use_mpi:
+
+            from schwimmbad import MPIPool
+
+            if self.is_master():
+                logger.info("... using Schwimmbad MPIPool")
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:469]: What happens if we don't add the master check here? Do we just get really noisy logging?

File: autofit/non_linear/parallel/sneaky.py [hunk 7]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+            An iterable to map over
+
+        """
+        return self.pool.map(function, iterable)
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            self.pool.terminate()
+        except:  # noqa
+            pass
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:503]: Either catch a more specific exception or log the exception. Silently catching all exceptions can lead to very difficult to find bugs.

File: autofit/non_linear/parallel/sneaky.py [hunk 8]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+        try:
+            del (FunctionCache.fitness, FunctionCache.prior_transform,
+                 FunctionCache.fitness_args, FunctionCache.fitness_kwargs,
+                 FunctionCache.prior_transform_args, FunctionCache.prior_transform_kwargs)
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:507]: Could these attributes have a default value of None so they did not have to be deleted? Or does the existence of the attributes play some important role in communicating state whilst multi processing?

File: autofit/non_linear/parallel/sneaky.py [hunk 9]
```diff
@@ -349,3 +350,171 @@ def __del__(self):
...
+        except:  # noqa
```
Comment (rhayes777) [inline autofit/non_linear/parallel/sneaky.py:508]: This should specifically catch an AttributeError

**Review by rhayes777 — APPROVED** [review]
I guess some of these comments are actually on changes introduced by @jacob-hjortlund 

Nothing major just a few pointers on best practices.

**Comment by Jammy2211** [comment]
I have not implemented the changes but will merge as I want everything on `main`.

Jacob's deadline is near so we can work these in at a later date!

**Pull request merged**